# CAPSTONE PROJECT

### Importing Necessary Libraries

In [24]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Obtaining the Data¶

Firstly, we need a list of neighbourhoods of Toronto. For the Toronto neighborhood data, a Wikipedia page exists :https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
BeautifulSoup and requests packages will be used for website scrapping. 

In [25]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
from bs4 import BeautifulSoup
soup = BeautifulSoup(website_url,'html.parser')
neighborhoods_data = soup.find('table',{'class':'wikitable sortable'})
neighborhoods_data_str = str(neighborhoods_data)
df = pd.read_html(neighborhoods_data_str)[0]
df = df.rename(columns = {'Postcode':'PostalCode','Neighbourhood':'Neighborhood'})
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

### Data Cleaning

There should be done 3 corrections in the dataset. 
1. In Brough column, the rows which are Not assigned will be removed.
2. Postal code areas which have more than one neighbourhood will be combined into one row.
3. In Neighbourhood column, the rows which are Not assigned will be changed to the brough name.

In [26]:
# Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
df.drop(df[df['Borough']=="Not assigned"].index,axis=0, inplace=True)
#More than one neighborhood can exist in one postal code area. 
df1 = df.groupby("PostalCode", as_index=False).agg(lambda x:','.join(set(x)))
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
df1.loc[df1['Neighborhood']=="Not assigned",'Neighborhood']=df1.loc[df1['Neighborhood']=="Not assigned",'Borough']
df1.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Malvern,Rouge
1        M1C  Scarborough  Rouge Hill,Port Union,Highland Creek
2        M1E  Scarborough       Morningside,West Hill,Guildwood
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [27]:
df1.shape

(103, 3)

The dataset has 103 rows and 3 columns.

In order to get venue data and plot a map, the coordinates of each neighnourhood is required. It is obtained from a csv file :http://cocl.us/Geospatial_data

In [28]:
# Read a file that includes coordinates of postal codes 
coordinates = pd.read_csv("/home/lpy/Downloads/Geospatial_Coordinates.csv")
coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [29]:
# Adding 'Latitude' and 'Longitude'columns to our dataframe
df1['Latitude']=coordinates['Latitude'].values
df1['Longitude']=coordinates['Longitude'].values
df1.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Malvern,Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill,Port Union,Highland Creek  43.784535   
2        M1E  Scarborough       Morningside,West Hill,Guildwood  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Since the project focuses on the Toronto only, the data will be filtered.

In [30]:
df2 = df1[df1['Borough'].str.contains("Toronto")].reset_index(drop=True)
df2.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

### Exploratory Data Analysis

#### Neighbourhoods of Toronto Map

geopy library is used to get the latitude and longitude values of Toronto.

In [31]:
address = 'Toronto'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [32]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df2['Latitude'], df2['Longitude'], df2['Borough'], df2['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Explore Neighborhoods in Toronto

Foursquare API will be used to explore venues in the neighbourhoods in Toronto. Top 100 venues which are within a radius of 1000 meters will be requested from Foursquare API.

In [47]:
# Define Foursquare Credentials and Version
CLIENT_ID = 'ULJAWACSBIL41ZFLQZJR4EO4P3KDHXFBIMGWHPVBSZIRC5L1' # your Foursquare ID
CLIENT_SECRET = 'SSZVJPBVK4FVV55M4WL42QWVOTPPKPGCHYIX0F0HCQVKS01L' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [48]:
#Create a function that gets the information for all neighbourhoods in Toronto
LIMIT = 100
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [49]:
# run the created function to get venues
toronto_venues = getNearbyVenues(names=df2['Neighborhood'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )


The Beaches
Riverdale,The Danforth West
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Rathnelly,Forest Hill SE,Summerhill West,Deer Park,South Hill
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Regent Park,Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
King,Richmond,Adelaide
Toronto Islands,Union Station,Harbourfront East
Toronto Dominion Centre,Design Exchange
Commerce Court,Victoria Hotel
Roselawn
Forest Hill West,Forest Hill North
Yorkville,North Midtown,The Annex
University of Toronto,Harbord
Chinatown,Kensington Market,Grange Park
CN Tower,Island airport,Railway Lands,King and Spadina,South Niagara,Bathurst Quay,Harbourfront West
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Exhibition Place,Parkdale Village,Brockton
The Junction South,High Park
Roncesvalles,Parkda

The size of resulting dataframe is as follows:

In [50]:
print(toronto_venues.shape)
toronto_venues.head()

(3083, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

               Venue  Venue Latitude  Venue Longitude  \
0    Tori's Bakeshop       43.672114       -79.290331   
1    The Fox Theatre       43.672801       -79.287272   
2     The Beech Tree       43.680493       -79.288846   
3    Ed's Real Scoop       43.672630       -79.287993   
4  Beaches Bake Shop       43.680363       -79.289692   

                  Venue Category  
0  Vegetarian / Vegan Restaurant  
1            Indie Movie Theater  
2                      Gastropub  
3                 Ice Cream Shop  
4                         Bakery

#### Analyze Each Neighbourhood

In [51]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols = toronto_onehot.columns.tolist()
cols.insert(0, cols.pop(cols.index('Neighborhood')))
toronto_onehot = toronto_onehot.reindex(columns = cols)


toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0  The Beaches                  0               0                  0        0   
1  The Beaches                  0               0                  0        0   
2  The Beaches                  0               0                  0        0   
3  The Beaches                  0               0                  0        0   
4  The Beaches                  0               0                  0        0   

   Airport Lounge  American Restaurant  Amphitheater  Animal Shelter  \
0               0                    0             0               0   
1               0                    0             0               0   
2               0                    0             0               0   
3               0                    0             0               0   
4               0                    0             0               0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Auto Garage  \
0                 0                   0                0            0   
1                 0                   0                0            0   
2                 0                   0                0            0   
3                 0                   0                0            0   
4                 0                   0                0            0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       1     0    0               0                 0   

   Basketball Stadium  Beach  Beach Bar  Beer Bar  Beer Store  \
0                   0      0          0         0           0   
1                   0      0          0         0           0   
2                   0      0          0         0           0   
3                   0      0          0         0           0   
4                   0      0          0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Bookstore  Boutique  \
0                   0          0       0           0          0         0   
1                   0          0       0           0          0         0   
2                   0          0       0           0          0         0   
3                   0          0       0           0          0         0   
4                   0          0       0           0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Station  \
0                0       0             0              0            0   
1                0       0             0              0            0   
2                0       0             0              0            0   
3                0       0             0              0            0   
4                0       0             0              0            0   

   Bus Stop  Business Servi

In [52]:
toronto_onehot.shape

(3083, 274)

Next, rows will be grouped by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [54]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Berczy Park                0.0   
1  Business Reply Mail Processing Centre 969 Eastern                0.0   
2  CN Tower,Island airport,Railway Lands,King and...                0.0   
3                         Cabbagetown,St. James Town                0.0   
4                                 Central Bay Street                0.0   

   Adult Boutique  Afghan Restaurant  Airport  Airport Lounge  \
0             0.0                0.0   0.0000          0.0000   
1             0.0                0.0   0.0000          0.0000   
2             0.0                0.0   0.0625          0.0625   
3             0.0                0.0   0.0000          0.0000   
4             0.0                0.0   0.0000          0.0000   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0             0.010000           0.0             0.0           0.0       0.0   
1             0.020833           0.0             0.0           0.0       0.0   
2             0.000000           0.0             0.0           0.0       0.0   
3             0.026316           0.0             0.0           0.0       0.0   
4             0.010000           0.0             0.0           0.0       0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0.02        0.00             0.000000               0.0   
1         0.00        0.00             0.020833               0.0   
2         0.00        0.00             0.000000               0.0   
3         0.00        0.00             0.000000               0.0   
4         0.01        0.01             0.030000               0.0   

   Athletics & Sports  Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  \
0                 0.0              0.0          0.0   0.010000        0.01   
1                 0.0              0.0          0.0   0.020833        0.00   
2                 0.0              0.0          0.0   0.000000        0.00   
3                 0.0              0.0          0.0   0.000000        0.00   
4                 0.0              0.0          0.0   0.000000        0.00   

     Bakery  Bank       Bar  Baseball Field  Baseball Stadium  \
0  0.030000   0.0  0.000000             0.0               0.0   
1  0.020833   0.0  0.020833             0.0               0.0   
2  0.000000   0.0  0.000000             0.0               0.0   
3  0.026316   0.0  0.000000             0.0               0.0   
4  0.010000   0.0  0.020000             0.0               0.0   

   Basketball Stadium     Beach  Beach Bar  Beer Bar  Beer Store  \
0                0.01  0.010000        0.0      0.04    0.000000   
1                0.00  0.020833        0.0      0.00    0.000000   
2                0.00  0.000000        0.0      0.00    0.000000   
3                0.00  0.000000        0.0      0.00    0.026316   
4                0.00  0.000000        0.0      0.01    0.000000   

   Belgian Restaurant  Bike Shop    Bistro  Board Shop  Bookstore  Boutique  \
0                0.01        0.0  0.010000         0.0       0.00       0.0   
1                0.00        0.0  0.020833         0.0       0.00       0.0   
2                0.00        0.0  0.000000         0.0       0.00       0.0   
3                0.00        0.0  0.000000         0.0       0.00       0.0   
4                0.00        0.0  0.000000         0.0       0.01       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0        0.020000   0.0000          0.0   
1                   0.0        0.020833   0.0625          0.0   
2                   0.0        0.000000   0.0000          0.0   
3                   0.0        0.000000   0.0000          0.0   
4                   0.0        0.020000   0.0000          0.0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Station  \
0             0.01     0.0      0.000000       0.000000          0.0   
1             0.00     

Since Coffee Shops are the interest of this project, the dataframe will be filtered to coffee shops only.

In [59]:
cof_shop = toronto_grouped[["Neighborhood","Coffee Shop"]]
cof_shop.head()

Neighborhood  Coffee Shop
0                                        Berczy Park     0.070000
1  Business Reply Mail Processing Centre 969 Eastern     0.083333
2  CN Tower,Island airport,Railway Lands,King and...     0.187500
3                         Cabbagetown,St. James Town     0.052632
4                                 Central Bay Street     0.080000

#### Cluster Neighborhoods

k-means clustering method will be used to analyze the data. The number of clusters, which is k, is determined to be 3. 

In [61]:
# set number of clusters
kclusters = 3

cof_shop_clustering = cof_shop.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(cof_shop_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 2, 0, 1, 2, 1, 1, 2, 2, 2], dtype=int32)

After obtaining the labels, they are added to the data.

In [62]:
# create a copy of coffee shop-neighborhood data
cof_shop_df = cof_shop.copy()
# add cluster labels
cof_shop_df["Cluster Label"] = kmeans.labels_

In [64]:
cof_shop_df.head()

Neighborhood  Coffee Shop  \
0                                        Berczy Park     0.070000   
1  Business Reply Mail Processing Centre 969 Eastern     0.083333   
2  CN Tower,Island airport,Railway Lands,King and...     0.187500   
3                         Cabbagetown,St. James Town     0.052632   
4                                 Central Bay Street     0.080000   

   Cluster Label  
0              1  
1              2  
2              0  
3              1  
4              2

In [72]:
# Add latitude/longitude for each neighborhood
coffee_shop = df2.join(cof_shop_df.set_index('Neighborhood'), on='Neighborhood')

coffee_shop.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Coffee Shop  Cluster Label  
0 -79.293031     0.060976              1  
1 -79.352188     0.090000              2  
2 -79.315572     0.054054              1  
3 -79.340923     0.070000              1  
4 -79.388790     0.111111              2

#### Visualization of Resulting Clusters

In [73]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(coffee_shop['Latitude'], coffee_shop['Longitude'], coffee_shop['Neighborhood'], coffee_shop['Cluster Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Clusters

Cluster 0

In [74]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 0]

PostalCode           Borough  \
5         M4P   Central Toronto   
10        M4W  Downtown Toronto   
13        M5A  Downtown Toronto   
27        M5V  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
5                                    Davisville North  43.712751 -79.390197   
10                                           Rosedale  43.679563 -79.377529   
13                           Regent Park,Harbourfront  43.654260 -79.360636   
27  CN Tower,Island airport,Railway Lands,King and...  43.628947 -79.394420   

    Coffee Shop  Cluster Label  
5      0.130000              0  
10     0.130435              0  
13     0.160000              0  
27     0.187500              0

In [78]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 0]['Coffee Shop'].mean()

0.15198369565217393

Cluster 1

In [75]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 1]

PostalCode           Borough  \
0         M4E      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
6         M4R   Central Toronto   
11        M4X  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   

                                       Neighborhood   Latitude  Longitude  \
0                                       The Beaches  43.676357 -79.293031   
2                     The Beaches West,India Bazaar  43.668999 -79.315572   
3                                   Studio District  43.659526 -79.340923   
6                                North Toronto West  43.715383 -79.405678   
11                       Cabbagetown,St. James Town  43.667967 -79.367675   
14                          Ryerson,Garden District  43.657162 -79.378937   
15                                   St. James Town  43.651494 -79.375418   
16                                      Berczy Park  43.644771 -79.373306   
18                           King,Richmond,Adelaide  43.650571 -79.384568   
19  Toronto Islands,Union Station,Harbourfront East  43.640816 -79.381752   
20          Toronto Dominion Centre,Design Exchange  43.647177 -79.381576   
23               Forest Hill West,Forest Hill North  43.696948 -79.411307   
24                Yorkville,North Midtown,The Annex  43.672710 -79.405678   
25                    University of Toronto,Harbord  43.662696 -79.400049   
26          Chinatown,Kensington Market,Grange Park  43.653206 -79.400049   
29            First Canadian Place,Underground city  43.648429 -79.382280   
30                                         Christie  43.669542 -79.422564   
32                          Little Portugal,Trinity  43.647927 -79.419750   
33       Exhibition Place,Parkdale Village,Brockton  43.636847 -79.428191   
34                     The Junction South,High Park  43.661608 -79.464763   
35                            Roncesvalles,Parkdale  43.648960 -79.456325   

    Coffee Shop  Cluster Label  
0      0.060976              1  
2      0.054054              1  
3      0.070000              1  
6      0.068182              1  
11     0.052632              1  
14     0.060000              1  
15     0.060000              1  
16     0.070000              1  
18     0.050000              1  
19     0.070000              1  
20     0.060000              1  
23     0.066667              1  
24     0.040000              1  
25     0.030000              1  
26     0.030000              1  
29     0.070000              1  
30     0.050000              1  
32     0.040000              1  
33     0.060000              1  
34     0.061224              1  
35     0.060000              1

In [79]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 1]['Coffee Shop'].mean()

0.05636829606675826

Cluster 2

In [76]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 2]

PostalCode           Borough  \
1         M4K      East Toronto   
4         M4N   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
12        M4Y  Downtown Toronto   
17        M5G  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
28        M5W  Downtown Toronto   
31        M6H      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
1                         Riverdale,The Danforth West  43.679557 -79.352188   
4                                       Lawrence Park  43.728020 -79.388790   
7                                          Davisville  43.704324 -79.388790   
8                          Moore Park,Summerhill East  43.689574 -79.383160   
9   Rathnelly,Forest Hill SE,Summerhill West,Deer ...  43.686412 -79.400049   
12                               Church and Wellesley  43.665860 -79.383160   
17                                 Central Bay Street  43.657952 -79.387383   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
36                                  Runnymede,Swansea  43.651571 -79.484450   
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   

    Coffee Shop  Cluster Label  
1      0.090000              2  
4      0.111111              2  
7      0.090000              2  
8      0.084746              2  
9      0.103896              2  
12     0.080000              2  
17     0.080000              2  
21     0.080000              2  
22     0.074074              2  
28     0.090000              2  
31     0.073529              2  
36     0.093333              2  
37     0.083333              2

In [80]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 2]['Coffee Shop'].mean()

0.08723254847880968

According to the cluster averages, Cluster 1 has the minimum frequency of occurrence of Coffee Shop. So that it is reasonable to make further exploration on those neighborhoods with the label cluster 1 since merket competition is low in those areas.

#### Examine Other Venues in Selected Cluster (Cluster 1)

In [83]:
# make a new dataframe called target that includes cluster 1 data
target = coffee_shop.loc[coffee_shop['Cluster Label'] == 1]
target.shape

(21, 7)

In [98]:
test =toronto_grouped.loc[toronto_grouped['Neighborhood'].isin(target['Neighborhood'])]
#target_neighborhood=toronto_grouped[toronto_grouped['Neighborhood']==target['Neighborhood']]
test.shape

(21, 274)